### ISSUE TITLE
OWIN-compatible implementation for "PerWebRequest" lifestyle

### ISSUE THREAD

**Issue by mario-d-s** [castleproject/Windsor#283]
The [PerWebRequest HttpModule]( Windsor/src/Castle.Windsor/MicroKernel/Lifestyle/PerWebRequestLifestyleModule.cs) has served everyone well for applications hosted in IIS, to enable "Per webrequest" lifestyle.

However, in recent years, OWIN has become more popular and with it, the ability to self-host websites. Windsor's PerWebRequestLifeStyle is incompatible with OWIN's self-hosting mechanism, as HTTP modules only work in the IIS pipeline.

Some OWIN-compatible implementations of this lifestyle have been hacked together, see [StackOverflow](https://stackoverflow.com/q/31412230/1313143) and [GitHub](https://github.com/BredStik/OWIN.Windsor.DependencyResolverScopeMiddleware).

However I think it would be best if such implementation was cleaned up, put under test, and integrated with the main project here. It would be a nice bonus if the same configuration syntax could be used as well (i.e. calling `LifestylePerWebRequest()`). Maybe there could be some magic to make it work under all hosting scenarios.

**Comment by jonorossi** [comment]
@mario-d-s completely agree the IIS module isn't useful for new web applications. There has been some work around ASP.NET Core and its built-in container in #120. I'd really like to see those with an interest in this to get together (probably via GitHub) to discuss what needs to happen and ultimately put forward those changes.

We should have a release of Windsor out with .NET Core support very soon, so I'll be great to have this type of support come in a following release.

**Comment by fir3pho3nixx** [comment]
@mario-d-s: +1 

PerWebRequest tries to implement dependencies like IHttpModule and then call back into the micro kernel using scopes, I think we should be inverting this. BeginScope/EndScope was the start of that work and was meant to supersede the other implementations but it was based on remoting which was not available in netcoreapp via netstandard(CallContexts) during the migration of Windsor. I do believe in netstandard is bringing this back in v2.0. 
 
Once it compiles everywhere, it would be a doddle to implement extensions in that wrap this idea.

**Comment by mario-d-s** [comment]
@jonorossi it's great to see that Windsor and Castle.Core are almost completely ready for .NET Core!
However, from what I've been reading on the web, there are a couple of differences between OWIN in .NET Core and OWIN in full framework. I have 0 experience with Core myself so I can't really comment.

I just hope any compatibility will also be backported to work with OWIN on the full framework.

@fir3pho3nixx According to at least one Microsoft engineer, CallContext should be avoided in scenario's unrelated to remoting. The source of that claim would be behind [this **broken** link](https://connect.microsoft.com/VisualStudio/feedback/details/276325/provide-better-context-support-to-async-operations) but unfortunately there seems to be no way of retrieving that page anymore (archive.org doesn't have it).

**Comment by jonorossi** [comment]
@fir3pho3nixx with the port we've left the .NET Framework code for scopes using `CallContext` (because we are targeting .NET 4.5), however the .NET Standard code will use `AsyncLocal`.

**Comment by jonorossi** [comment]
> I just hope any compatibility will also be backported to work with OWIN on the full framework.

@mario-d-s this is where I highly encourage you to get involved in the project, we really do need help and features won't get supported unless someone actually wants them and puts in some work. I hope as a collective updates to the per-webrequest lifestyle can support everyone involved.

**Comment by fir3pho3nixx** [comment]
@mario-d-s: I remember reading about this many moons ago, cannot for the life of me remember why though :)

@jonorossi: I did come across the use of AsyncLocal [here](https://github.com/volosoft/castle-windsor-ms-adapter/blob/master/src/Castle.Windsor.MsDependencyInjection/MsLifeTimeScope.cs#L40). I must have missed that bit in the netcore branch when my internet was borked. 

Checking out some middleware implementations would be as simple as creating an extension that goes something like this: 

```csharp 

app.Use(new Func<AppFunc, AppFunc>(next => (async env =>
{
    myWindsorContainer.BeginScope();
    await next.Invoke(env);
    myWindsorContainer.EndScope()
})));

```

You could then wrap it in a tidy extension method.

**Comment by mario-d-s** [comment]
@jonorossi I would love to get involved, and it's something I will definitely begin to do in the near future when I get around to finally learn Git 😃 

@fir3pho3nixx Correct me if I'm wrong, but such middleware would use the default scope (i.e. `LifeTimeScoped()`). What if you need scopes in other places. Does this support nesting?

**Comment by fir3pho3nixx** [comment]
> Correct me if I'm wrong, but such middleware would use the default scope (i.e. LifeTimeScoped())

@mario-d-s: I could not find the particular extension `LifeTimeScoped` but I am pretty sure you mean `LifestyleScoped`. This is basically a shorthand API extension that will apply LifestyleType.Scoped to the config for that registration(using descriptors) which eventually ends up creating a ScopedLifestyleManager(commit: 4e7716e2) in the DefaultKernel. This is significant because when it tries to obtain the current scope for an instance it will return a CallContextLifetimeScope(the class based on remoting I mentioned earlier written by Kozmic himself).  The default scope you mention(`DefaultLifetimeScope`) was only really a thing in the `PerWebRequestLifestyleModule` and the `ThreadScopeAccessor`. 

> What if you need scopes in other places.

@mario-d-s: There is an overload for LifestyleScoped whereby you can specify the type of an accessor. This is a concept which allows you get these things from anywhere but you have to implement an `IScopeAccessor`. Pretty neat huh?

> Does this support nesting?

@mario-d-s: From what I can tell it does. 

I think there was significant work at play here to try and clean this up and to make lifestyle management a little easier in general and if Windsor doesn't have it you could "roll your own" in a very easy way, so I am glad you raised this issue. I like the public API for BeginScope/EndScope because of it's simplicity, I am not particularly bound to the implementation of it, I just mentioned the current implementation to add more info to the discussion.

**Comment by fir3pho3nixx** [comment]
> @jonorossi I would love to get involved, and it's something I will definitely begin to do in the near future when I get around to finally learn Git 😃

Would be awesome if you got involved, there is a community that could field any questions you might have. If you are learning Git @jonorossi is a boss with it, he improved my Git-Fu 😃

**Comment by fir3pho3nixx** [comment]
@mario-d-s - I was thinking we start a series of PR's to start moving the DesktopCLR `PerWebRequestLifestyleModuleRegistration` stuff out of Windsor and into facilities(or something else). It is still there and supports the [FEATURE_SYSTEM_WEB ](https://github.com/castleproject/Windsor/tree/netcore). The reason I say this is there has been a significant effort to migrate this windsor to dotnet core and it prevents different challenges for Windsor's public API when it comes to lifestyle management. My vote is that we talk about how we would abstract this out whilst implementing an OWIN compatible version at the same time? What do you think?

**Comment by fir3pho3nixx** [comment]
@jonorossi - I would like to start work on this next. Can you tell me what you would like to see happen here?

My recommendation is we start creating satellite assemblies(Windsor Lifestyle NuGet's) for vendor specific stuff and start breaking it out of the MicroKernel. Our versioning can then be kept in step with vendors and not be tied to core implementations of Windsor. Perhaps we want to start leveraging the facility architecture which would be keeping with the spirit of the original design or we could just create them as extension methods(honestly I prefer this approach). 

 - Identify how we want to get PerWebRequest out of the MicroKernel.
 - Extend it to OWIN in a separate release. 

Look forward to your feedback.

**Comment by jonorossi** [comment]
A new assembly sounds like a great way to go, that keeps the dependencies of `Castle.Windsor.dll` small and keeps things self contained. I see no reason the `System.Web` lifestyle couldn't also move out of the main assembly, which also helps by removing some conditional compilation. The lifestyle enum should be reserved for the really basic built-in lifestyles, and really its use de-emphasised.

Extension methods are definitely a good approach, take a look at the `Castle.Windsor.Lifestyles` contrib project that did the same thing: https://github.com/castleproject-deprecated/Castle.Windsor.Lifestyles/blob/master/Castle.Windsor.Lifestyles/LifestyleRegistrationExtensions.cs

**Comment by fir3pho3nixx** [comment]
Yes, this is what I was thinking. Can I go ahead and create a new branch called `lifestyles`?

Also, do we want a separate assembly for each target implementation? eg.

 - Castle.Windsor.Lifestyles.WebRequest
 - Castle.Windsor.Lifestyles.OWIN
 - Castle.Windsor.Lifestyles.NetCore

If you are happy with this as a starting block I can add these new projects and their co-respective NuGet packing requirements into the VS2017 build infrastructure as empty projects for now. I will then commence work on moving `PerWebRequest` down.

**Comment by jonorossi** [comment]
> Can I go ahead and create a new branch called lifestyles?

Go for it.

> Also, do we want a separate assembly for each target implementation?

That probably makes sense as the package dependencies would be vastly different, I'd probably use these names instead:

- Castle.Windsor.Lifestyles.~~WebRequest~~ SystemWeb
- Castle.Windsor.Lifestyles.~~OWIN~~ Owin 
- Castle.Windsor.Lifestyles.~~NetCore~~ AspNetCore (I think I'm wrong on this one, maybe .MSDependencyInjection instead)

Unless they'll be more than just a lifestyle, i.e. the WCF integration has a per channel lifestyle but is obviously named a facility as the main part is the Windsor facility.

> I will then commence work on moving PerWebRequest down.

Feel free to do what you think works best, but I'd have thought it would be best to leave that one until you've got the others the way you want them.

**Comment by fir3pho3nixx** [comment]
I have done some more investigation into this. My interest here was how we `future proof` Castle Windsor and create an implementation that services both dotnet core and desktop clr using OWIN because it is already available on both.  

All paths appear to be leading to Rome, or at least in this case `Microsoft.Extensions.DependencyInjection`. I did know that this could be used on netcore but did not for one second think about using it in a desktop clr scenario. A bit of a `duh` moment I must admit. 

My example is only using the `Microsoft.Extensions.DependencyInjection` nuget sans any Windsor container implementation. The reason I have done this is because I want to understand how the dependencies and the consumer API works before we dive into `Microsoft.Extensions.DependencyInjection.Abstractions`, which is where the meat and potatoes of the cross runtime implementation might need to be implemented if we all agree on my suggested approach. 

I submit my cobbled together OWIN [startup file](https://github.com/fir3pho3nixx/castle-windsor-owin-deps/blob/master/OwinWeb/Startup.cs) for `net452`. Where you can see the same old divergent IDependencyResolver implementation for MVC [here](https://github.com/fir3pho3nixx/castle-windsor-owin-deps/blob/master/OwinWeb/Startup.cs#L22) and WebApi [here](https://github.com/fir3pho3nixx/castle-windsor-owin-deps/blob/master/OwinWeb/Startup.cs#L25). I then went ahead and registered my own service called [IAnyService](https://github.com/fir3pho3nixx/castle-windsor-owin-deps/blob/master/OwinWeb/Startup.cs#L31). Following that I registered an [MVC](https://github.com/fir3pho3nixx/castle-windsor-owin-deps/blob/master/OwinWeb/Startup.cs#L32) and a [WebApi](https://github.com/fir3pho3nixx/castle-windsor-owin-deps/blob/master/OwinWeb/Startup.cs#L33) controller. 

This got me round to thinking about how we still use Windsor's registration API but still get the buttery goodness of integration using MVC and WebAPI in a DesktopClr/NetCore scenario. The answer appears to be by re-implementing `Microsoft.Extensions.DependencyInjection.Abstractions`.  I then had a look at the competition and it appears as though they have beat us to the [punch](https://github.com/autofac/Autofac.Extensions.DependencyInjection). This could be viewed as a good thing because they have solved the problem for us. We just need a Windsor flavour.

@mario-d-s: Can you weigh in on your intended usage of Castle.Windsor for this issue? How you would like this API to work? I am thinking this [kind of extension](https://github.com/autofac/Autofac.Extensions.DependencyInjection/blob/develop/src/Autofac.Extensions.DependencyInjection/ServiceCollectionExtensions.cs#L42) is perfectly inline with the spirit of registering OWIN middleware(`Use*` convention). My suggestion here is we have a `UseCastleWindsor` extension. I would also like to know whether you have any apprehension netstandard being in as a dependency? Assuming you are using this in a DesktopClr scenario, this can slow down CI builds. Are there also any other `Service Location` concerns you think we are missing whilst registering middleware(seen some stackoverflow's on this)? 

@jonorossi: I think the test project for this should boot up my example as a OWIN self host, use the windsor registration API for `IAnyService` which is ultimately resolved in [MVC](https://github.com/fir3pho3nixx/castle-windsor-owin-deps/blob/master/OwinWeb/Controllers/HomeController.cs#L10) and [WebApi](https://github.com/fir3pho3nixx/castle-windsor-owin-deps/blob/master/OwinWeb/Controllers/ValuesController.cs#L11) using a HttpClient. Would be super cool to get a dotnet core version of this going too. It is supported.

Let me know what you guys think.

**Comment by mario-d-s** [comment]
@fir3pho3nixx I think you're pretty much on track for this. I would definitely refer to the AutoFac stuff as an inspiration to get started.

I don't have any additional concerns for the moment. I can't really comment on your question about the `netstandard` dependency, I assume this is .NET Core stuff which, I've already mentioned, I have no experience with.

By the way, I've already learned a bit of Git in the last couple of weeks. Perhaps I'll be able to contribute myself in a meaningful way here soon, I'll keep you posted.

**Comment by jonorossi** [comment]
I agree with @mario-d-s, it sounds like you are on the right track.

> I think the test project for this should boot up my example as a OWIN self host, use the windsor registration API for IAnyService which is ultimately resolved in MVC and WebApi using a HttpClient. Would be super cool to get a dotnet core version of this going too. It is supported.

I've got no objection to adding some integration tests, it might be worthwhile having a unit test project just for the MSDependencyInjection project though so anyone working on core Windsor doesn't have to worry about all of that and its dependencies.

**Comment by fir3pho3nixx** [comment]
@mario-d-s - I am sorry if this is not relevant to your issue, which I have already solved [BTW](https://github.com/fir3pho3nixx/castle-windsor-owin-deps/tree/fuk-c/OwinWeb.IoC). I do have an interest in getting dotnet core going for the `Microsoft.Extensions.DependencyInjection` stuff. If you feel this is not relevant here I am happy to take a slap on the hand and raise a new issue. I just feel that the implementation is incredibly naive and vulnerable to to memory leaks when it comes to scoping. I submit further detail to @jonorossi below. 

@jonorossi - Can you weigh in on this(without pulling your punches)? `Microsoft.Extensions.DependencyInjection` makes some bold assumptions about naked parent level transients that are resolved in a scoped lifestyle. 

This [test](https://github.com/fir3pho3nixx/castle-windsor-owin-deps/blob/fuk-c/OwinAspNetCore.IoC.Tests/DependencyInjectionSpecificationTests.cs#L412) right here assumes that `[scoped](https://github.com/fir3pho3nixx/castle-windsor-owin-deps/blob/fuk-c/OwinAspNetCore.IoC.Tests/DependencyInjectionSpecificationTests.cs#L428)` lifestyles will auto-magically [dispose](https://github.com/fir3pho3nixx/castle-windsor-owin-deps/blob/fuk-c/OwinAspNetCore.IoC.Tests/DependencyInjectionSpecificationTests.cs#L442) [transients](https://github.com/fir3pho3nixx/castle-windsor-owin-deps/blob/fuk-c/OwinAspNetCore.IoC.Tests/DependencyInjectionSpecificationTests.cs#L431) without being housed in a parent dependency that is marked as `scoped`. I got this test from [here](https://github.com/aspnet/DependencyInjection/issues/552#issuecomment-317274810) and copied it into the solution for debugging purposes. This to me is quite simply bad design. It is non deterministic and opens up the gates of hell for prematurely disposed transients or alternatively could create further memory leaks if the developer changing this code makes an assumption that his/her transient object is running within a scope. There simply is no type safety around this design. 

I have also been reviewing [custom lifestyles](https://github.com/volosoft/castle-windsor-ms-adapter/blob/master/src/Castle.Windsor.MsDependencyInjection/MsScopedLifestyleManager.cs) over [here](https://github.com/volosoft/castle-windsor-ms-adapter) and was wondering if this was truly the answer. It feels like a sledge hammer.

My expectation for this test would be something like [this](https://github.com/fir3pho3nixx/castle-windsor-owin-deps/blob/fuk-c/CastleWindsorConsole/Program.cs#L68) where by you could find usages of [OptionsContainer](https://github.com/fir3pho3nixx/castle-windsor-owin-deps/blob/fuk-c/CastleWindsorConsole/Program.cs#L76) and immediately know it is scoped. Better no?

**Comment by fir3pho3nixx** [comment]
I am almost tempted to create a parent type around transients that is a mixin for scoped lifestyles.

**Comment by fir3pho3nixx** [comment]
I am tempted to wrap transients using Castle Core mixins with fallback logic. Not sure if this would work, but I think I am going to start looking at this tomorrow night.

**Comment by fir3pho3nixx** [comment]
PS: The numerous registrations of duplicate types are not helping. I am not impressed. This stuff is garbage as far as I am concerned.

**Comment by jonorossi** [comment]
> This test right here assumes that scoped lifestyles will auto-magically dispose transients without being housed in a parent dependency that is marked as scoped. I got this test from here and copied it into the solution for debugging purposes. This to me is quite simply bad design. It is non deterministic and opens up the gates of hell for prematurely disposed transients or alternatively could create further memory leaks if the developer changing this code makes an assumption that his/her transient object is running within a scope. There simply is no type safety around this design.

How does Windsor behave today resolving and disposing transients as the root inside a scope, does the burden cause it to get released on scope disposal?

**Comment by jonorossi** [comment]
> I am tempted to wrap transients using Castle Core mixins with fallback logic. Not sure if this would work, but I think I am going to start looking at this tomorrow night.

The performance penalty would be quite high especially for all transients. How would this work anyway, how would you know the user is finished with the object, I can't see how you'd do reference counting.

The transient really should be released as per Windsor's long standing guidelines.

**Comment by fir3pho3nixx** [comment]
>How does Windsor behave today resolving and disposing transients as the root inside a scope, does the burden cause it to get released on scope disposal?

No it does not, and I don't believe it should. If it is resolved as a child dependency where the parent has a scope it does. I think this is correct behaviour.  

>The performance penalty would be quite high especially for all transients. How would this work anyway, how would you know the user is finished with the object, I can't see how you'd do reference counting.

Yeah, thinking about it I don't think it would be easy and anything as best would be a lifecycle hack. I will keep digging but I might raise an issue to see what the authors say. Lifestyle `side effects` like this can lead to all sorts of bad things.

**Comment by fir3pho3nixx** [comment]
Yeah not sure how to proceed, raised this [issue](https://github.com/aspnet/DependencyInjection/issues/558). I think this needs fixing. Parent/Child scoping got violated in Windsor when I implemented [this](https://github.com/fir3pho3nixx/castle-windsor-owin-deps/tree/fuk-d/OwinAspNetCore.IoC/TransientWithScopeSideEffects). Let's see what happens next.

**Comment by fir3pho3nixx** [comment]
The [duplicate registrations](https://github.com/aspnet/DependencyInjection/issues/559) of `RC1ForwardingActivator` are also doing my head in.

**Comment by fir3pho3nixx** [comment]
Getting tests going was a little [tricky](https://github.com/aspnet/DependencyInjection/issues/560) also.

**Comment by jonorossi** [comment]
Looking at that test code again it feels a little like Windsor's child containers since the resolution of all components is happening against the scope, but then the singleton doesn't get disposed on scope disposal so it doesn't really match. When the code references `scope.ServiceProvider` is that the exact same instance as the `provider` variable or something special for that specific scope? Just wondering if nested scopes are a thing. If they are the same, the code would read clearer with just `provider.GetService`.

If the transient gets disposed at the end of the scope then why shouldn't the singleton also get disposed? Is there a design document/specification for how the MSDependencyInjection abstractions should be implemented and how the lifecycle of components behave with scopes?

	ServiceCollectionServiceExtensions.AddSingleton<IFakeSingletonService, FakeService>(collection);
	ServiceCollectionServiceExtensions.AddScoped<IFakeScopedService, FakeService>(collection);
	ServiceCollectionServiceExtensions.AddTransient<IFakeService, FakeService>(collection);
	...

	using (var scope = ServiceProviderServiceExtensions.CreateScope(provider))
	{
		disposableService = (FakeService) scope.ServiceProvider.GetService<IFakeScopedService>();
		transient1 = (FakeService) scope.ServiceProvider.GetService<IFakeService>();
		transient2 = (FakeService) scope.ServiceProvider.GetService<IFakeService>();
		singleton = (FakeService) scope.ServiceProvider.GetService<IFakeSingletonService>();

		Assert.False(disposableService.Disposed);
		Assert.False(transient1.Disposed);
		Assert.False(transient2.Disposed);
		Assert.False(singleton.Disposed);
	}

	Assert.True(disposableService.Disposed);
	Assert.True(transient1.Disposed);
	Assert.True(transient2.Disposed);
	Assert.False(singleton.Disposed);

**Comment by fir3pho3nixx** [comment]
>Looking at that test code again it feels a little like Windsor's child containers since the resolution of all components is happening against the scope, but then the singleton doesn't get disposed on scope disposal so it doesn't really match. 

We appear to have matching behaviour for singletons. I am thinking the reasoning behind this not getting disposed is because although it is a memory leak it is a small one because of the long lived lifestyle. 

>When the code references scope.ServiceProvider is that the exact same instance as the provider variable or something special for that specific scope? Just wondering if nested scopes are a thing. If they are the same, the code would read clearer with just provider.GetService

In the case of their recommended test (which uses xunit) I have debugged the `IServiceProviderFactory` which would suggest that the instances could be different for scope.ServiceProvider but it is the same instance. So for now I am in complete agreement with you. 

>If the transient gets disposed at the end of the scope then why shouldn't the singleton also get disposed?

They ask you not to used scoped dependencies in a singleton in the docs. So I guess this is just something the user has to remember. Then again thinking about it, if you use a transient in a singleton according this test it will get disposed but the singleton will live on ....

>Is there a design document/specification for how the MSDependencyInjection abstractions should be implemented and how the lifecycle of components behave with scopes?

This test suite is what they came back with from GitHub. The [docs online](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/dependency-injection) point you to an autofac implementation. It is a long old read.

**Comment by fir3pho3nixx** [comment]
They just got back to me, BeginScope() = new WindsorContainer(). So will go ahead and try this.

**Comment by fir3pho3nixx** [comment]
Right, after blasting @davidfowl with WTF's he pointed me to https://github.com/aspnet/Mvc/issues/5403. 

Apparently there is an implementation that use an outside-in approach for containers [here](https://github.com/dotnetjunkie/Missing-Core-DI-Extensions/blob/master/src/SampleApplication.CastleWindsor/Startup.cs) by bastardising scopes using OWIN middleware. I am definitely not saying this is the answer, it needs to be validated but luckily we are not the only ones. I am going to drop this business of trying to write my own abstraction implementation and join this discussion. I just need time to absorb what is going on here. I resent the fact that registration is [fault tolerant](https://github.com/aspnet/DataProtection/commit/4dad47eeab6be84c70d431a21fdc4d152ca332ea#diff-42452680fb1b858efb10577a09d0ae26R33), I also dont like the fact that service resolution is done from [scopes](https://github.com/aspnet/DependencyInjection/blob/dev/src/Microsoft.Extensions.DependencyInjection.Abstractions/IServiceScope.cs#L19) with overriding lifestyles when it comes to disposables. This is a radical departure from windsor's design. Scopes are something else entirely. Will keep this issue updated.

**Comment by jonorossi** [comment]
@fir3pho3nixx looking forward to hearing the outcome of your continued research and discussions, I wasn't aware of those non-conforming GitHub issues either.

**Comment by jonorossi** [comment]
Knowing that at least the ASP.NET Core DependencyInjection [implementation is less about a lifestyle and more an adapter](https://github.com/volosoft/castle-windsor-ms-adapter/issues/17), we should instead follow the existing [Windsor facilities](https://github.com/castleproject/Windsor/blob/master/docs/facilities.md) naming convention used for all existing extensions.

- `Castle.Facilities.SystemWeb`
- `Castle.Facilities.Owin`
- `Castle.Facilities.MsDependencyInjection`

**Comment by fir3pho3nixx** [comment]
@jonorossi - I have to be honest here, I believe we should drop the whole `MsDependencyInjection` thing for this issue. It is massive, not to be forgotten though, will raise a new issue. On the bright side I have solved the OWIN implementation on both desktop clr and core. A massive huge shout out to @dotnetjunkie from the CastleProject for coming up with tidy working a working OWIN implementation [here](https://github.com/dotnetjunkie/Missing-Core-DI-Extensions/blob/master/src/SampleApplication.CastleWindsor/Startup.cs).  You are a legend. From all my time wasted studying this the only change I needed to make was implement [naming](https://github.com/fir3pho3nixx/castle-windsor-owin-deps/blob/master/OwinAspNetCoreOwin/Startup.cs#L94), and mark things as [IsDefault](https://github.com/fir3pho3nixx/castle-windsor-owin-deps/blob/master/OwinAspNetCoreOwin/Startup.cs#L93). Both WebAPI and MVC now work. This supports my scoping API argument earlier in this issue. Good stuff. 

Still need to write a hammer test and check it with scenarios I believe are valid using dottrace and dotmemory. Still more work to be done before I can submit a PR. I also need to add sustainable testing.

**Comment by dotnetjunkie** [comment]
> They ask you not to used scoped dependencies in a singleton in the docs. So I guess this is just something the user has to remember.

This is actually something they have "addressed" post-v1. There is now an overload of `ServiceCollectionContainerBuilderExtensions.BuildServiceProvider` that accepts a boolean `validateScopes` that prevents scoped instances to be resolved from the root container (and probably prevent them from being injected into singletons).

The problem however with this new feature is that 1. Nobody knowns about its existence and 2. It is something the user must explicitly call and 3. setting this is a breaking change which might cause 3rd party libraries that integrate with the `IServiceCollection` to break and 4 using it is rather ugly. Example:

``` c#
public IServiceProvider ConfigureServices(IServiceCollection services)
{
    // your normal stuff here

    // Explictly call BuildServiceProvider and return the IServiceProvider
    return services.BuildServiceProvider(validateScopes: true);
}

```

> Then again thinking about it, if you use a transient in a singleton according this test it will get disposed but the singleton will live on ....

Great catch. By allowing transients to be injected into singletons, they basically use the lifestyle that Autofac calls 'instance per dependency'. In other words, "each consumer is intended to get its own instance of the dependency, and the dependency is expected to live as long as the consumer". Whether or not this is sane is arguable, but if you do this, you simply can't dispose any transient that is injected into a singleton, because disposing the transient means you break its consuming singleton and therefore break the application. Long story short: These two models don't marry.

>  he pointed me to aspnet/Mvc#5403.

He forgot to point out to other discussions such as this important one: https://github.com/aspnet/DependencyInjection/pull/416 and this one https://github.com/aspnet/DependencyInjection/issues/433. I've been [discussing](https://github.com/simpleinjector/SimpleInjector/issues/41#issuecomment-110770132) this with Microsoft for about two years now, and have written a two articles on this myself [here](https://simpleinjector.org/blog/2016/06/whats-wrong-with-the-asp-net-core-di-abstraction/) and [here](https://simpleinjector.org/blog/2016/07/working-around-the-asp-net-core-di-abstraction/) over a year ago.

Great that Fowler points at #5403, but up till now, Microsoft has done nothing to improve the situation for non-conformers.

**Comment by jonorossi** [comment]
> I have to be honest here, I believe we should drop the whole MsDependencyInjection thing for this issue. It is massive, not to be forgotten though, will raise a new issue. On the bright side I have solved the OWIN implementation on both desktop clr and core.

@fir3pho3nixx sounds good to me.

> He forgot to point out to other discussions such as this important one. I've been discussing this with Microsoft for about two years now, and have written a two articles on this myself here and here over a year ago. Great that Fowler points at #5403, but up till now, Microsoft has done nothing to improve the situation for non-conformers.

@dotnetjunkie thanks for the links and the great work blazing the trails on this stuff over the last few years. After skim reading a few of those it is coming back to me, I was following along back then in the early DNX days but just checked out completely because it all seemed too hard and I didn't have the time or energy to care at the time.

**Comment by fir3pho3nixx** [comment]
@jonorossi - Current learnings so far. 

### OWIN has some constraints: 

 - It supports AspNet (MVC & WebApi) only in IIS (`web host`) on DesktopCLR. 

 - It only supports WebApi `self host` on DesktopCLR. 

 - It is not a named thing on AspNetCore but they did take the idea of [middleware](https://github.com/fir3pho3nixx/Windsor/blob/d74ed47969c71462c689397b95cec99b483150a6/src/Castle.Facilities.Owin/KestrelHost/RequestScopeMiddlewareInstaller.cs#L44) with them and re-used the concept of [Startup.cs](https://github.com/fir3pho3nixx/Castle.Facilities.Owin.TestHarness/blob/master/AspNetCoreApp/Startup.cs) files. They also have some tidy options for customisation ([IStartupFilter](https://github.com/fir3pho3nixx/Windsor/blob/d74ed47969c71462c689397b95cec99b483150a6/src/Castle.Facilities.Owin/KestrelHost/RequestScopeMiddlewareInstaller.cs#L23), [IControllerActivator](https://github.com/fir3pho3nixx/Windsor/blob/d74ed47969c71462c689397b95cec99b483150a6/src/Castle.Facilities.Owin/KestrelHost/Activators/ControllerActivator.cs#L23), ...).

 - Kestrel(Linux/Windows) and WebListener(Windows only) are the hosts for AspNetCore. 

My PR addresses the DesktopCLR side so far with some peculiar API's namely `UseWindsorOwinWebHost` and `UseWindsorOwinSelfHost` (needs renaming though!). I am working on the `UseWindsorOwinKestrel` implementation at the moment. 

This got me round to thinking, we do have an opportunity for creating a migration safe API given everything I know about how Windsor runs across the spectrum. 

### Considerations for how we name assemblies

Most people would not want to have MVC and WebApi in the same project as a dependency in a DesktopCLR scenario. The reason I say this, is because `PerWebRequest` does it using a module that requires config after finding a runtime exception. The newer `zero-config` way unfortunately requires 2 separate projects.  

**MVC**

```csharp
DependencyResolver.SetResolver(new WebHost.MsMvcDependencyResolver());
```

requires(latest)

```xml
<PackageReference Include="Microsoft.AspNet.Mvc" Version="5.2.3" />
```

**WebApi**

```csharp
GlobalConfiguration.Configuration.DependencyResolver = new MsWebApiDependencyResolver<WebHost.MsWebApiDependencyScope>();
```

requires(latest)

```xml
<PackageReference Include="Microsoft.AspNet.WebApi" Version="5.2.3" />
```

In both WebHost and SelfHost, the API's above have nothing to do with OWIN for overriding from a Windsor point of view but as such it might be prudent to herd all the environment config sheep([App_Start](https://github.com/fir3pho3nixx/Castle.Facilities.Owin.TestHarness/tree/master/AspNetWebHost/App_Start)) into a StartUp.cs file that is OWIN compliant. Also now that we know Startup.cs and middleware is a thing on netcore with completely different dependencies perhaps there is an opportunity for API optimisation.  

### AspNet was the only thing tying this all together in a fly by the seat of your pants kind of way

I assume you identified `Castle.Facilities.SystemWeb` as the landing place for `PerWebRequest` lifestyles, maybe it needs a `web` bucket. For example we could tweak this to be `Castle.Facilities.Web.SystemWeb` so we can support `Castle.Facilities.Web.Mvc` or `Castle.Facilities.Web.WebApi` as an example. Also in the interest of hiving off functionality that should not be in the microkernel, perhaps we should make it `Castle.Facilities.MicroKernel.Web.PerWebRequest`? Not sure myself at this point. 


### Proposed new `facility` namespace Castle.Facilities.AspNet.*

With this we can create: 

 - Castle.Facilities.AspNet.SystemWeb

This contains a push down of `PerRequest` API as an extension method of microkernel lifestyles ... 

 - Castle.Facilities.AspNet.Mvc


This contains a `AddWindsorMvc` extension on [Owin.IAppBuilder](https://msdn.microsoft.com/en-us/library/microsoft.owin.builder.appbuilder(v=vs.113).aspx) but override's the [MVC ControllerFactory](https://github.com/fir3pho3nixx/Windsor/blob/facilities-netcore/src/Castle.Facilities.Owin/WindsorOwinExtensions.cs#L40).

 - Castle.Facilities.AspNet.WebApi

This contains a `AddWindsorWebApi` extension on [Owin.IAppBuilder](https://msdn.microsoft.com/en-us/library/microsoft.owin.builder.appbuilder(v=vs.113).aspx) but override's the [WebApi DependencyResolver](https://github.com/fir3pho3nixx/Windsor/blob/facilities-netcore/src/Castle.Facilities.Owin/WindsorOwinExtensions.cs#L50) with this [scope](https://github.com/fir3pho3nixx/Windsor/blob/facilities-netcore/src/Castle.Facilities.Owin/WebHost/MsWebApiDependencyScope.cs).

 - Castle.Facilities.AspNet.WebApi.SelfHost

This contains a `AddWindsorWebApiSelfHost` extension on [Owin.IAppBuilder](https://msdn.microsoft.com/en-us/library/microsoft.owin.builder.appbuilder(v=vs.113).aspx) but overrides the [WebApi DependencyResolver](https://github.com/fir3pho3nixx/Windsor/blob/facilities-netcore/src/Castle.Facilities.Owin/WindsorOwinExtensions.cs#L65) with this [scope](https://github.com/fir3pho3nixx/Windsor/blob/facilities-netcore/src/Castle.Facilities.Owin/SelfHost/MsWebApiDependencyScope.cs)

 - Castle.Facilities.AspNetCore.Kestrel

This contains a `AddWindsorKestrel` extension [yet to be verified](https://www.google.co.uk/search?q=picture+cat&rlz=1C1JZAP_enGB719GB719&source=lnms&tbm=isch&sa=X&ved=0ahUKEwjJ5cyF8bbVAhWhJ8AKHd67CqQQ_AUICigB&biw=1920&bih=950).

 - Castle.Facilities.AspNetCore.WebListener

This contains a `AddWindsorWebListener` extension [yet to be verified](https://www.google.co.uk/search?q=picture+cat&rlz=1C1JZAP_enGB719GB719&source=lnms&tbm=isch&sa=X&ved=0ahUKEwjJ5cyF8bbVAhWhJ8AKHd67CqQQ_AUICigB&biw=1920&bih=950).

**Comment by jonorossi** [comment]
@fir3pho3nixx I was really hoping others would have jumped in with their opinion, if you are following along silently please do.

I do like the `Castle.Facilities.AspNet.*` naming compared to adding `Web` to the namespace to `Ms` prefixes everywhere. The product name (ASP.NET or ASP.NET Core) makes it much clearer what it is for. I just wonder if all of this is necessary since we don't have any today, and ASP.NET is now sort of legacy since you can run ASP.NET Core on .NET Framework and .NET Core?

It would also be worthwhile thinking how the naming would tie into usage of Windsor in Nancy (or another web framework) on .NET Core using `DependencyInjection`.

**Comment by masterpoi** [comment]
What would be the difference between AddWindsorKestrel  and AddWindsorWebListener? As i understand it the request scope would be opened and closed in middleware, which afaik behaves the same in both? Since AspNetCore middleware is not Owin middleware, I'd rather suggest something like Castle.Facilities.AspNet.Core and Castle.Facilties.Owin.

**Comment by fir3pho3nixx** [comment]
>I do like the Castle.Facilities.AspNet.* naming compared to adding Web to the namespace to Ms prefixes everywhere. The product name (ASP.NET or ASP.NET Core) makes it much clearer what it is for. I just wonder if all of this is necessary since we don't have any today, and ASP.NET is now sort of legacy since you can run ASP.NET Core on .NET Framework and .NET Core?

@jonorossi - Glad you like the product namespaced idea. As for ASPNET sort of being legacy, there are a surprising amount of companies in the center of London that still have not made any progress on moving over to ASPNET Core(let alone the cloud!).  Selfishly I am also tired of picking up projects in the wild with half-baked Windsor solutions in them.  I do take your point, though most these integrations appear to be about 2 lines of code. Something worth considering. Perhaps we should look forward not back. That said I will leave the final decision to you. :)

>It would also be worthwhile thinking how the naming would tie into usage of Windsor in Nancy (or another web framework) on .NET Core using DependencyInjection. 

Nancy yes, DependencyInjection potentially but I would encourage use of the Windsor registration API instead. The [Nancy Windsor Bootstrapper](https://github.com/NancyFx/Nancy.Bootstrappers.Windsor) could benefit from an update. It appears to be using the same DesktopClr method the Microkernel uses for PerWebRequest lifestyles and only runs under [net452](https://github.com/NancyFx/Nancy.Bootstrappers.Windsor/blob/master/src/Nancy.Bootstrappers.Windsor/Nancy.Bootstrappers.Windsor.csproj#L12). I can factor this into my thinking whilst checking out the core stuff further. Good shout.  

>What would be the difference between AddWindsorKestrel and AddWindsorWebListener? As i understand it the request scope would be opened and closed in middleware, which afaik behaves the same in both? Since AspNetCore middleware is not Owin middleware, I'd rather suggest something like Castle.Facilities.AspNet.Core and Castle.Facilties.Owin.

@masterpoi - Nothing at all from what I can tell. These are just methods I spit balled out there, they can be revised to only be `AddWindsor`. Castle.Facilities.Owin is an interesting one because of the WebHost(that can do true PerWebRequest lifestyles hosted in IIS) and then there is the pure middleware logic for SelfHost(which is vanilla CallContextLifetimeScope without handlers and HttpContext.Current == null). I would like to know just for clarity, would Castle.Facilities.Owin only support WebApi?

**Comment by masterpoi** [comment]
> Castle.Facilities.Owin is an interesting one because of the WebHost(that can do true PerWebRequest lifestyles hosted in IIS) and then there is the pure middleware logic for SelfHost(which is vanilla CallContextLifetimeScope without handlers and HttpContext.Current == null). I would like to know just for clarity, would Castle.Facilities.Owin only support WebApi?

(Just some thoughts, my understanding of the details or consequences might be lacking.)

Wouldn't this be something that you could auto-detect? In fact, when reading the OP again, i think putting all this magic under a single LifestylePerWebRequest is the way to go. Decouple the framework from the lifestyle (also with regard to naming). You would want to be able to use PerWebRequest lifestyle outside of the established frameworks, i.e. in custom middleware.

So LifestylePerWebRequest would match (if i understood correctly), on
IIS: Traditional PerWebRequest
OWIN - IIS: Traditional PerWebRequest
OWIN - Selfhost : CallContextLifetimeScope
AspNetCore: LifeStyleScoped

In the case of AspNetCore you would call AddWinsor somewhere to register the middleware that manages the scope.

In addition to that you would have the facilities/adapters that provide integration with Mvc, WebApi, Nancy, MsDependencyInjection for Core, and other frameworks.

**Comment by fir3pho3nixx** [comment]
>(Just some thoughts, my understanding of the details or consequences might be lacking.)
No problem, I am only connecting the dots as I go. Glad to have your input. 

>Wouldn't this be something that you could auto-detect?

Yes, this was my first thought whilst pursuing this under the [SelfHost](https://github.com/fir3pho3nixx/Castle.Facilities.Owin.TestHarness/blob/master/AspNetWebApiSelfHost/Startup.cs#L46) OWIN implementation on [net452](https://github.com/fir3pho3nixx/Castle.Facilities.Owin.TestHarness/blob/master/AspNetWebApiSelfHost/AspNetWebApiSelfHost.csproj#L11) here.  

My question here would be this, if you were testing for the presence of an HttpContext.Current in an OWIN Startup how would you build that auto detection mechanism without incurring a performance hit in the middleware along with having to make it threadsafe? 

>In fact, when reading the OP again, i think putting all this magic under a single LifestylePerWebRequest is the way to go. 

My worry here is coupling. The Windsor project already houses a reference to System.Web because of LifestylePerWebRequest in the MicroKernel. This could be improved, for the WPF guys who dont care about this. With everything I have explained so far in this issue we have to try decouple the following packages for DesktopClr on AspNet. 

```xml
<package id="Microsoft.AspNet.Mvc" version="5.2.3" targetFramework="net452" />
<package id="Microsoft.AspNet.WebApi" version="5.2.3" targetFramework="net452" />
<package id="Microsoft.AspNet.WebApi.WebHost" version="5.2.3" targetFramework="net452" />
<package id="Microsoft.AspNet.WebApi.SelfHost" version="5.2.3" targetFramework="net452" />
```

For core we are lucky all of these concepts are merged. It does bring new problems. 

>Decouple the framework from the lifestyle (also with regard to naming)

I am open to suggestions. How would you achieve this?

>You would want to be able to use PerWebRequest lifestyle outside of the established frameworks, i.e. in custom middleware.

```
So LifestylePerWebRequest would match (if i understood correctly), on
IIS: Traditional PerWebRequest
OWIN - IIS: Traditional PerWebRequest
OWIN - Selfhost : CallContextLifetimeScope
AspNetCore: LifeStyleScoped
```

I am looking at the runtime level. Here is my view incorporating yours.

* DesktopClr (net452)
  * WebHost(IIS)
    * AspNet.Mvc: (PerWebRequest becomes LifeStyleScoped -> CallContextLifetimeScope -> Remoting)
    * AspNet.WebApi: (PerWebRequest becomes LifeStyleScoped -> CallContextLifetimeScope -> Remoting)
  * SelfHost
    * AspNet.WebApi: (LifeStyleScoped -> CallContextLifetimeScope -> Remoting)
* NetCore (netcoreapp1.1)
  * Kestrel/WebListener
    * AspNetCore: (LifeStyleScoped -> CallContextLifetimeScope -> AsyncLocal)

My goal here is we agree on how a side effect like System.Web found it's way into the Lifestyle namespace of the MicroKernal in Windsor? This is an important API but it should be scaled out to a new assembly and it's runtime implications considered. BeginScope/refDispose looks like a lovely API to do this with when it comes to LifeStyleScoped because it way more portable across platforms right now. Please see [this](https://github.com/fir3pho3nixx/Castle.Facilities.Owin.TestHarness/blob/master/AspNetCoreApp/Startup.cs) which I shamelessly plagiarized from @dotnetjunkie. The are really tidy hooks outward from the MicroKernel which would make the assembly Castle.Windsor almost `runtime` agnostic.  

> In the case of AspNetCore you would call AddWinsor somewhere to register the middleware that manages the scope.

Yes. I have played with this idea [here](https://github.com/fir3pho3nixx/Castle.Facilities.Owin.TestHarness/tree/master/AspNetCoreApp).

> In addition to that you would have the facilities/adapters that provide integration with Mvc, WebApi, Nancy, MsDependencyInjection for Core, and other frameworks.

Other framework perhaps not, I am being selfish here. I have web concerns. Not sure what people would need from the Desktop world. Nobody is looking at this right now.

**Comment by masterpoi** [comment]
> My goal here is we agree on how a side effect like System.Web found it's way into the Lifestyle namespace of the MicroKernal in Windsor? This is an important API but it should be scaled out to a new assembly and it's runtime implications considered. BeginScope/refDispose looks like a lovely API to do this with when it comes to LifeStyleScoped because it way more portable across platforms right now. Please see this which I shamelessly plagiarized from @dotnetjunkie. These are really tidy hooks outward from the MicroKernel which would make the assembly Castle.Windsor almost runtime agnostic.

Yes, this sounds really great. Make everything lifestylescoped and move the reference to System.Web to a seperate assembly for those who need it.

Since everything would become LifestyleScoped, have you not de facto decoupled the framework from the lifestyle? Facilities providing the deeper integration would then configure how to setup LifestyleScoped for the specific framework. 

For instance, the Castle.Facilities.AspNetCore does the middleware thing as shown in your example and provide adapters for MsDependencyInjection (which, i guess, should be in another assembly for those cases where you use MsDependencyInjection outside of asp.net)

Castle.Facilities.AspNet.Mvc would provide the ControllerFactory, DependencyResolver, etc...

**Comment by fir3pho3nixx** [comment]
@masterpoi - Sorry I have been busy with other things, I am just pinging you back to say hi and I am still thinking about this. 

> Yes, this sounds really great. Make everything lifestylescoped and move the reference to System.Web to a seperate assembly for those who need it.

Glad you agree. lifestylescoped is the intellectual property of castle windsor, and facilities implement this for different platforms. 

> Since everything would become LifestyleScoped, have you not de facto decoupled the framework from the lifestyle?

Yes. The lifestyle is really down to the hosting environment, they know best.

> Facilities providing the deeper integration would then configure how to setup LifestyleScoped for the specific framework.

Correct. 

> For instance, the Castle.Facilities.AspNetCore does the middleware thing as shown in your example and provide adapters for MsDependencyInjection (which, i guess, should be in another assembly for those cases where you use MsDependencyInjection outside of asp.net)

Maybe we leave MsDependencyInjection out of it for now. I don't think it's scoping is compatible with the MicroKernel because of the way it expects to create and destroy the container the whole time. My [proposal](https://github.com/castleproject/Windsor/issues/283#issuecomment-319493389) is purely runtime based

**Comment by fir3pho3nixx** [comment]
I think we are nearing the home run on this issue. I have implemented 3 facilities one of which supports OWIN self hosting.

**Comment by fir3pho3nixx** [comment]
@mario-d-s - if you could please take a look at https://github.com/fir3pho3nixx/Windsor/tree/facilities-desktopclr and tell me if this is good for your use case. Any last minute changes would be good. 

The docs are located here: https://github.com/fir3pho3nixx/Windsor/blob/facilities-desktopclr/docs/aspnetwebapi-facility.md

Any feedback would be much appreciated.

**Comment by mario-d-s** [comment]
@fir3pho3nixx I just had a look at the docs you've written, but I'm afraid I'm missing the point of this facility entirely. You still register controllers as transient, but using the facility makes them magically "per web request"?

As a matter of fact, we already had custom IHttpControllerActivator and IDependencyResolver implementations as the perfect place to plug Windsor into the ASP.NET framework and start resolving controllers from the container. That is I guess how everyone uses Windsor in an ASP.NET project, I can't think of another way to make it possible.

The OWIN (IIS-hosted) project I'm working on now has transient controllers, resolved through that custom IDependencyResolver implementation and it works as it should, I'm not even sure if the old PerWebRequest lifestyle would still have any value when it comes to controllers...

Aside from that, I guess PerWebRequestLifeStyle had another functionality: scope other services (not controllers) to web requests implicitly. I don't believe we use that anywhere in our projects but I don't see how the facility you wrote supports that?

I'm not criticizing here by the way, I just feel like I'm missing something.

I'll see if I find some time this week to pull in your repository, make a build and play around a bit.

**Comment by fir3pho3nixx** [comment]
> @fir3pho3nixx I just had a look at the docs you've written, but I'm afraid I'm missing the point of this facility entirely. You still register controllers as transient, but using the facility makes them magically "per web request"?

Transients would be scoped to the lifestyle of the request based on the resolve/release policy. Transients consumed multiple times as dependents of controllers though would behave as expected and then be tracked as burdens with decomission concerns. Do you believe this is incorrect?

**Comment by fir3pho3nixx** [comment]
> Aside from that, I guess PerWebRequestLifeStyle had another functionality: scope other services (not controllers) to web requests implicitly. I don't believe we use that anywhere in our projects but I don't see how the facility you wrote supports that?

Yes, we moved PerWebRequest to it's own facility SystemWeb. You implicitly get PerWebRequest using the `LifeStyleScoped` method on the facility. I thought this was what we were saying in this discussion. Feel free to tell me if you thought otherwise.

**Comment by fir3pho3nixx** [comment]
> I'm not criticizing here by the way, I just feel like I'm missing something.

Nope not at all, I really value your feedback :)

**Comment by mario-d-s** [comment]
> Transients would be scoped to the lifestyle of the request based on the resolve/release policy. Transients consumed multiple times as dependents of controllers though would behave as expected and then be tracked as burdens with decomission concerns. Do you believe this is incorrect?

Technically that is correct, but since controllers should be the composition root (comparable to `Program.Main()` in a console app) I believe it's terrible design to have them as dependencies in other components. Still, you are right on that part and it would be the main difference compared to the transient lifestyle.

> Yes, we moved PerWebRequest to it's own facility SystemWeb. You implicitly get PerWebRequest using the LifeStyleScoped method on the facility. I thought this was what we were saying in this discussion. Feel free to tell me if you thought otherwise.

That is indeed what the discussion was about. However, I thought `LifeStyleScoped` works when you are using `Container.BeginScope()` calls. Unless you have a custom scope accessor, but I can't find that in your docs or commits. With the original PerWebRequest lifestyle, one did not need to do `BeginScope()` or similar calls, it was fully implicit.

**Comment by fir3pho3nixx** [comment]
> Oops, total misinterpretation there. If I understand correctly, using the facility, all transients would instead become per-web-request? But what if you really needed a transient, how can you still register it?

You still have transients, it just they behave differently at the controller level. You have to remember that they will get resolved and released using the activator which only lasts as long as the request. In your dependency graph went 3 levels deep and the same transient service was consumed inside of that I would expect them to be 2 completely different instances.

**Comment by fir3pho3nixx** [comment]
> However, I thought LifeStyleScoped works when you are using Container.BeginScope() calls. Unless you have a custom scope accessor, but I can't find that in your docs or commits

You can see the BeginScope/EndScope logic [here](https://github.com/fir3pho3nixx/Windsor/commit/a54f6f4334ec8e50f0cc67ef22fe1f157c13c753#diff-4e7d930e415981230e002e70b1d67d9cR97).

**Comment by fir3pho3nixx** [comment]
>  With the original PerWebRequest lifestyle, one did not need to do BeginScope() or similar calls, it was fully implicit.

You are effectively getting this if you register your controllers as either transient or scoped. The difference is that with scope's you need to opt in using `WithScopedLifestyles`. You can find this in the docs over [here](https://github.com/fir3pho3nixx/Windsor/commit/a54f6f4334ec8e50f0cc67ef22fe1f157c13c753#diff-b56aa016acadd2a734d5f5d390cc3562R44)

**Comment by fir3pho3nixx** [comment]
PS: Thanks for this, I will also update the docs to describe the lifestyles better. If there is anything you would like to ask or perhaps is not clear let me know and I will make some changes where ever I can to make the UX for this a little easier.

**Comment by mario-d-s** [comment]
> If your dependency graph went 3 levels deep and the same transient service was consumed inside of that twice I would expect them to be 2 completely different instances.

From the viewpoint of resolving components, that totally makes sense. Of course, dependencies of a controller will be per-web-request implicitly (the same way transient dependencies of a singleton are effectively singleton in that context). But in the following situation:

- Controller A depends on services M and N
- M and N depend on X and Y
- M, N are transient
- X is per web request, Y is transient

How do you register them so that M and N both get the same instance of X per request, but each get a different instance of Y per request?

I don't see how your proposed registration API makes that clear, since you are registering everything as transient?

> You are effectively getting this if you register your controllers as either transient or scoped. The difference is that with scope's you need to opt in using `WithScopedLifestyles`.

I guess I'm confused over how scopes work in Windsor, I've never really used them either. But I know the docs and the docs say basically:

- register as `LifeStyleScoped()` and use `Container.BeginScope()` + `scope.Dispose()`
- register as `LifeStyleScoped<CustomScopeAccessor>()`.

AFAIK you haven't implemented the `CustomScopeAccessor` bit for the per-web-request lifestyle here?

**Comment by fir3pho3nixx** [comment]
> How do you register them so that M and N both get the same instance of X per request, but each get a different instance of Y per request?

Make X scoped.

**Comment by fir3pho3nixx** [comment]
> I don't see how your proposed registration API makes that clear, since you are registering everything as transient?

Should I update the docs to Scoped?

**Comment by fir3pho3nixx** [comment]
> AFAIK you haven't implemented the CustomScopeAccessor bit for the per-web-request lifestyle here?

Don't need the scope accessor. It falls back on to CallContext's.  I did need it for SystemWeb facility though.

**Comment by fir3pho3nixx** [comment]
To make scopes a little clearer, you use a scope accessor when you want to tie the scope to some external thing. In the case of SystemWeb we use a HttpModule for this which is tied to EndRequests for scope disposal. So it make sense to use an accessor. 

In SelfHost we are falling back on to callcontext, which shares the scope for the current executing thread in a fancy way without carrying the same problems as using a ThreadStatic. 

Hope this makes sense.

**Comment by fir3pho3nixx** [comment]
@mario-d-s - I am just making sure we are covering the bases here, please refer to [this reply](https://github.com/castleproject/Windsor/issues/283#issuecomment-321077653) to @masterpoi in this discussion thread where I clarified how scopes would work. 

@masterpoi - A review from you is also most welcome.

**Comment by mario-d-s** [comment]
@fir3pho3nixx I've been digging into the code in Windsor to find how this whole idea of scopes works, and it's finally beginning to click!

Can you confirm I'm getting it please:

- In Windsor, `Scoped` components cannot be resolved without beginning a scope using e.g. `Container.BeginScope()`, otherwise you get an exception
- Your facility, if configured using `WithScopedLifestyles()`, begins a scope when a controller is resolved and ends it when a controller is released. This should happen only once per request, right?
- All services, including controllers, registered as `transient` are still `transient`, but you can make them `scoped` as a replacement of `per-web-request`. That only works if you have called `WithScopedLifestyles()`, otherwise you'll get exceptions that no scope is available.
- Controllers, although transient, are only released and resolved per-request by your `AspNetWebApiControllerActivator`, effectively making their lifestyle `per-web-request`. Which does make total sense of course.

If my reasonig is right, my only issue is that as a registration API, `LifestylePerWebRequest()` was really obvious anc clear what the lifestyle was going to be. A user of Windsor did not need to care how it worked under the hood. By now making the lifestyle scoped, you have to remember that a scope is being created and that it is really a per-web-request scope.

If possible, and if you agree, I would suggest you keep the `LifestyleScoped()` but instead use a custom scope accessor. I guess all you'd need to do is move a little bit of code from your facility to a scope accessor.

The registration API can then be `LifestyleScoped<WebRequestScopeAccessor>()` or similar, and you could even have another method `LifestylePerWebRequest()` that just calls the other one. Maybe it's a way to avoid breaking changes in the public API around this?

**Comment by mario-d-s** [comment]
There is one extra improvement I'd like to see, or at least a decision I'd like to understand better: you've chosen to still use the CallContext to store scopes into. I'm not super clear on how this works exactly, but is there a reason you can't bind it to some context that is closer to the request? I see you're [already using it to store the scope](https://github.com/fir3pho3nixx/Windsor/commit/a54f6f4334ec8e50f0cc67ef22fe1f157c13c753#diff-4e7d930e415981230e002e70b1d67d9cR102), so what's the difference?

**Comment by fir3pho3nixx** [comment]
> In Windsor, Scoped components cannot be resolved without beginning a scope using e.g. Container.BeginScope(), otherwise you get an exception

Correct

> Your facility, if configured using WithScopedLifestyles(), begins a scope when a controller is resolved and ends it when a controller is released. This should happen only once per request, right?

Correct again

> All services, including controllers, registered as transient are still transient, but you can make them scoped as a replacement of per-web-request. That only works if you have called WithScopedLifestyles(), otherwise you'll get exceptions that no scope is available.

In a self host this is done for you but in the web host scenario you definitely would have to opt in. In fact I actually try and wrap [the exception](https://github.com/castleproject/Windsor/pull/351/commits/a54f6f4334ec8e50f0cc67ef22fe1f157c13c753#diff-534b9974a3ec4023cdfa21c870793077R47) in the facility here to give you a more relevant self help message.

> Controllers, although transient, are only released and resolved per-request by your AspNetWebApiControllerActivator, effectively making their lifestyle per-web-request. Which does make total sense of course.

Correct

> If my reasonig is right, my only issue is that as a registration API, LifestylePerWebRequest() was really obvious anc clear what the lifestyle was going to be. A user of Windsor did not need to care how it worked under the hood. By now making the lifestyle scoped, you have to remember that a scope is being created and that it is really a per-web-request scope.

A scope for me is the more general use case with some really extensible API's for how they are accessed. I also thought they had been around for a while. If you check the first commit for the [ScopedLifestyleManager](https://github.com/castleproject/Windsor/commits/aa9b8b353ee2e533d586495eec254e216f800c09/src/Castle.Windsor/MicroKernel/Lifestyle/ScopedLifestyleManager.cs) you will notice it goes all the way back 2011 which was over 5 years ago. I am also not sure I follow you when you say people will just have to remember. I my hope is that the documentation is clearly stating this, which I pointed out with a link earlier :)

> If possible, and if you agree, I would suggest you keep the LifestyleScoped() but instead use a custom scope accessor. I guess all you'd need to do is move a little bit of code from your facility to a scope accessor.

Why the scope accessor here? You can already do this with vanilla Windsor if you want to manager your own activator/dependency resolver. This facility is trying to help manage that for you, so you dont have to spent time googling the right way to achieve that. Also what other kind of scoped access would people want to support in an OWIN scenario? Are you perhaps thinking along the lines of middleware?

> The registration API can then be LifestyleScoped<WebRequestScopeAccessor>() or similar, and you could even have another method LifestylePerWebRequest() that just calls the other one. Maybe it's a way to avoid breaking changes in the public API around this?

3 things here: 

 - I still don't understand why you need the scope accessor. Where are you accessing the scope from? This really does not make sense to me right now because I am not using the HttpContext for WebApi.  

 - I like you're suggestion of bringing back `PerWebRequest` just not sure whether a namesake for familiarity sake to obscure what is really going on is a good idea. I will defer to @jonorossi on this approach. If you really want PerWebRequest it is very much alive and well(as it ever was) in the SystemWeb facility.

 - It would still be a breaking change because the namespaces would be different. That is why we are potentially targeting V5 for this stuff.

**Comment by fir3pho3nixx** [comment]
> There is one extra improvement I'd like to see, or at least a decision I'd like to understand better: you've chosen to still use the CallContext to store scopes into. I'm not super clear on how this works exactly, but is there a reason you can't bind it to some context that is closer to the request? 

This was designed in 2011 by @kkozmic but here is a link to [CallContext](https://msdn.microsoft.com/en-us/library/system.runtime.remoting.messaging.callcontext(v=vs.110).aspx).

**Comment by fir3pho3nixx** [comment]
@mario-d-s - You feedback has been great, thank you. Please see commit's where I have taken your input on board.

My only criticism for this issue is that you did not follow the discussion fully and I had to explain how scopes worked. In future it would be great if you limit the conscious streaming in terms of questions, and commit to comments like [this one](https://github.com/castleproject/Windsor/issues/283#issuecomment-308425688) where you pledged to commit code once you learned git. 

Please remember this is an open source project. If you raise issues, you should be prepared to get involved and commit code. 

Again I thank you for you're feedback but @jonorossi  and I will take it from here.

**Comment by mario-d-s** [comment]
@fir3pho3nixx I understand, I realize this was not the best place to seek further understanding about the topics I've asked about.

I still genuinely intend to get involved more, but this particular issue was too complicated for me either way. I've followed the discussions from the sideline though. I'm glad I was at least of some help.

The update you made to the docs on your facility makes it much clearer by the way.

Thanks for all the help!

**Comment by masterpoi** [comment]
@fir3pho3nixx I still think your proposed solutions is great.

 I took a quick look at the docs you wrote. Look good.

Just a quick thought i had, typically i'd not make my controllers scoped, but the database session / uow only. Making the controllers scoped might be a bit confusing as an example.
 
In the MVC case, the AddControllerAssembly method, i assume would scan the assembly for controllers? (And that WithLifestyleScopedPerWebRequest has nothing to do with the scanning?) Maybe name it ScanControllerAssembly instead and put WithLifestyleScopedPerWebRequest  in front ?

**Comment by fir3pho3nixx** [comment]
> Just a quick thought i had, typically i'd not make my controllers scoped, but the database session / uow only. Making the controllers scoped might be a bit confusing as an example.

Not all. Your UoW pattern hopefully breaks things down to a factory type which yields instance based resources which are disposable at some point. Could you post a Gist link to your high level UoW pattern? Maybe something like [this](https://docs.microsoft.com/en-us/aspnet/mvc/overview/older-versions/getting-started-with-ef-5-using-mvc-4/implementing-the-repository-and-unit-of-work-patterns-in-an-asp-net-mvc-application)?

> In the MVC case, the AddControllerAssembly method, i assume would scan the assembly for controllers? 

No. I intentionally avoided ComponentRegistration scenario's. I would really love it if people started using [IWindsorInstallers](https://github.com/castleproject/Windsor/blob/master/docs/installers.md) with [conventional registration](https://github.com/castleproject/Windsor/blob/master/docs/registering-components-by-conventions.md) properly. The facilities should become the application of those scopes. 

> And that WithLifestyleScopedPerWebRequest has nothing to do with the scanning?

Yes. For now.

> Maybe name it ScanControllerAssembly instead and put WithLifestyleScopedPerWebRequest in front ?

LifestyleScopedControllerSourceAssembly. New name welcome.

I guess this is where we can work some magic, to expose new ComponentRegistration extensions and see if we can do something to sweeten the registration part. It would make these facilities very handy. 

Thoughts/Gist?

**Comment by jonorossi** [comment]
@fir3pho3nixx I've reviewed #351.

**Comment by masterpoi** [comment]
> Not all. Your UoW pattern hopefully breaks things down to a factory type which yields instance based resources which are disposable at some point. Could you post a Gist link to your high level UoW pattern? Maybe something like this(interfaces only :))?

Yes, typically a factory method that provides a uow something. I.e. ravendb
```csharp
var ds = new DocumentStore() { /* settings */ };
container.Register(Component.For<IDocumentSession>().LifestyleScoped().UsingFactoryMethod(() => ds.OpenSession()));
```

In my mind everything that doesn't keep state should either be transient or maybe singleton for perf. Since controllers classes typically don't keep state, i tend to make them transient. (You typically won't need multiple controllers per request, so making them transient means in practice they have the same lifetime as making them scoped). The uow / db session may be shared between multiple services that the controller uses, so that really needs to be scoped to the request.

As i said, it might be confusing to some, but it might only be confusing to me ;)

 >> In the MVC case, the AddControllerAssembly method, i assume would scan the assembly for controllers?

> No.  I intentionally avoided ComponentRegistration scenario's. I would really love it if people started using IWindsorInstallers with conventional registration properly. The facilities should become the application of scopes.

Agreed, but then I guess i'm missing something. What does AddControllerAssembly do?

**Comment by fir3pho3nixx** [comment]
AddControllerAssembly merely supplies assembly to source types from when creating controllers for MVC. It has been converted to a fluent interface, you can now have many. 

Not sure about the UoW scenario, will have to get back to you.

**Comment by fir3pho3nixx** [comment]
@masterpoi  - I have done more digging here. 

Your example:

```csharp
var ds = new DocumentStore() { /* settings */ };
container.Register(Component.For<IDocumentSession>().LifestyleScoped().UsingFactoryMethod(() => ds.OpenSession()));
```  

I believe we can support this but please let's explore this a bit more. 

`->` means `has a dependency on`.

If you are using something like `ControllerType` -> `ServiceType` -> `RepositoryType` -> `IDocumentSession` then there are a few things to consider about lifestyles here when it comes to how things are resolved. 

You mentioned either transient or singleton lifestyle would be good:

> In my mind everything that doesn't keep state should either be transient or maybe singleton for perf

I agree with you completely. If the paradigm is a layered to hell architecture like in some of the awful implementations I have seen round London recently then things change ever so subtly depending on where they are resolved.

Transient. This one is tricky. You would always expect a new instance. However, if you mark a controller as transient and the facility only instantiates 1 when it engages the ASP.NET request pipeline via ControllerFactory or whatever the abstraction is. In this case you essentially have a "per web request" component right? Transient by the very definition is `PerWebRequest` scoped because of where it is consumed.

It get's more complicated when you move over to `RepositoryType`'s. N number of services resolve into the controller they consume repositories, but unfortunately services can consume other services and repositories again and this is where the difference comes in. 

Imagine this example:

Posts have comments, facebook right? You cannot add a comment without knowing what the original the post id was.

```
CommentController -> CommentService -> CommentRepository(instanceOf(1))
                                                                  -> PostService -> CommentRepository(instanceOf(2))
```

The net effect of transient being marked on the CommentController  vs the CommentRepository can yield different behaviours when it comes to state purely because of the way it is consumed. If CommentRepository was an in `memory instance` based repository it would almost always be wrong in this scenario.

What we are trying to say here, is that `scoped` is a thing(it combats these idiosyncrasies). We are also saying you need to consider the how it is consumed. Scoped in this case would fix the `state problem` of transients that hold N+1 upstream dependencies minus framework consumers because we cant control that(singletons should violate this law). 

Let's consider disposables for transients given the above. This where it can get a little crazy especially if you are using TypedFactories that them self re-engage the resolution mechanism of Windsor by using the (k) -> k.Resolve<> feature pattern. Burdens (tracking of disposables from what I can tell) simply wont allow this in their current state because they don't care about release policies or late bound factory components. 

What this means in plain English is that typed factories will hurt you bad, when it comes to what you "think is possible" and then you drop a "typed factory" in and it does not work(ref: https://github.com/castleproject/Windsor/pull/373).  

Moving on, so "unit of work". Please explain what your context is? I would not use this word lightly. In fact I find it quite offensive given the implementations I have seen in the wild because most people do not get what this `really` means. It is one of the most mis-interpreted patterns I have found across the realms of insurance, banking and e-commerce. I think it means transactional state with ACID and managing resources in a federated service oriented architecture. I would love to hear your experiences on this. 

That said, I have investigated using AutoFac with [Instance per matching lifetime scope](http://autofaccn.readthedocs.io/en/latest/lifetime/instance-scope.html#instance-per-matching-lifetime-scope). I honestly don't associate this with "unit of work" because it quite simply does not match the problem domain we are trying to solve in windsor. The taxonomy is completely wrong for a container. We can offer named scopes, similar such. When it is resolved that name needs to be supplied to the resolve mechanism. This would create  named scoped slices that could conform to lifestyle variants that would eventually support a unit of work(which we should have to care about in Windsor). The problem I have with all of this, it is supported today in Windsor. You just need to structure you code correctly.

Now this might seem like an attack, it is not. Please share what you are trying to do and with further discussion hopefully we can figure this out :)

**Comment by fir3pho3nixx** [comment]
@masterpoi - thanks for you input so far, it is a great help.

**Comment by masterpoi** [comment]
> I believe we can support this but please let's explore this a bit more.

What do you mean with support? I'm sorry if i let you believe otherwise, but i don't have any fundamental issues with how things are now. And again, the solution you are working on looks great.

The rest of my remarks can be summarized as "Let the architect choose the right lifestyles depending on his architecture." ;)

> Transient. This one is tricky. You would always expect a new instance. However, if you mark a controller as transient and the facility only instantiates 1 when it engages the ASP.NET request pipeline via ControllerFactory or whatever the abstraction is. In this case you essentially have a "per web request" component right? Transient by the very definition is PerWebRequest scoped because of where it is consumed.

Yes exactly.

> The net effect of transient being marked on the CommentController vs the CommentRepository can yield different behaviours when it comes to state purely because of the way it is consumed. If CommentRepository was an in memory instance based repository it would almost always be wrong in this scenario.

This is true if CommentRepository was also registered as Transient. But if it needs to keep state I would register it scoped. In general Repositories would need to be Scoped because they depend on the UOW.

I don't see how registering the controller transient has any impact here. If the controller would somehow need to be resolved again in the chain i think there is something really wrong in the design. Given that the controller is the interface between the framework (MVC) and the "user code", it should be managed by the framework. And hence, as said before, in this case Transient implies PerWebRequest.
And the controller being transient does not imply it's dependencies should be.

> What we are trying to say here, is that scoped is a thing(it combats these idiosyncrasies). We are also saying you need to consider the how it is consumed. Scoped in this case would fix the state problem of transients that hold N+1 upstream dependencies minus framework consumers because we cant control that(singletons should violate this law).

Agreed that scoped can fix this. But I really think you should consider every service in your dependency chain and give it the appropriate lifestyle when registering. Making the lifestyle of the dependencies hang of the controller lifestyle makes it harder to reason about them in those cases where you need the dependencies out of the scope of the controller. Making higher up dependencies transient gives flexibility for lower dependencies to be something more strict like scoped or singleton. The reverse is not true. If the controller is scoped,  every dependency is de facto also scoped. 

So yes, making the controllers scoped fixes the issue in your example. But i feel like it is better to link the lifestyle to the specific component demanding it.

> This where it can get a little crazy especially if you are using TypedFactories that them self re-engage the resolution mechanism of Windsor by using the (k) -> k.Resolve<> feature pattern.

Do you mean making Controllers transient would make it hard/impossible for Windsor to known when to dispose it's IDisposable dependencies? Yes, i can see this might become in issue in some designs, and in that case, the architect/designer has the choice to make the controllers Scoped for this reason. But i really consider that to be his job.

> I think it means transactional state with ACID and managing resources in a federated service oriented architecture. I would love to hear your experiences on this.

In this case i use uow to refer to the transaction boundary around database operations in a single process (i.e. inside a single (micro)service). I.e. the DbContext in EF, the DocumentSession in RavenDb / Marten, NServiceBus messagehandler.

> That said, I have investigated using AutoFac with Instance per matching lifetime scope. I honestly don't associate this with "unit of work" because it quite simply does not match the problem domain we are trying to solve in windsor. The taxonomy is completely wrong for a container. We can offer named scopes, similar such. When it is resolved that name needs to be supplied to the resolve mechanism. This would create named scoped slices that could conform to lifestyle variants that would eventually support a unit of work(which we should have to care about in Windsor). The problem I have with all of this, it is supported today in Windsor. You just need to structure you code correctly.

Agreed.

**Comment by fir3pho3nixx** [comment]
> What do you mean with support? I'm sorry if i let you believe otherwise, but i don't have any fundamental issues with how things are now. And again, the solution you are working on looks great.

Add some public API's to solve problems we *think* might be an issue. Don't ever say sorry, I do enough of that on project already :)

> The rest of my remarks can be summarized as "Let the architect choose the right lifestyles depending on his architecture." ;)

Read [this](https://medium.com/@JonMcLeodEA/enterprise-architecture-is-dead-33dd0e63cbbf). Feel free to rant but don't use names.

> So yes, making the controllers scoped fixes the issue in your example. But i feel like it is better to link the lifestyle to the specific component demanding it.

Interesting, can you elaborate on this. We have several problems with typed factories in this space at the moment. There is an issue where some big brains are going to duke this out. I am going to back off and see what happens. Not linking for cleanliness and separation of problems. Please give me your account. 

> Do you mean making Controllers transient would make it hard/impossible for Windsor to known when to dispose it's IDisposable dependencies?

No, deeper, chained resolves via vanilla windsor kernel that dependends on typed factories(using LateBoundComponents) with disposables at the end are a pain in the arse. It only crops up with N+2 resolution chains where a typed factory is in the middle. Hopefully we will have an answer soon. 

It is not often I get a reflective opinion that leans towards my own. Thanks for your input. We are nearing the home run on this issue. Closing this out as I have prepped the final PR. Release is imminent.

**Comment by fir3pho3nixx** [comment]
@hammett don't ever tell me my contributions are not worth it. I am not going anywhere.
